Added missing control packets: CW, DROPREQ, PEERERROR (#79)
Haivision/srt-rfc: draft-sharabayko-mops-srt.md

: A single value or a range of lost packets sequence numbers. See packet sequence number
coding in {{packet-seq-list-coding}}.

### Congestion Warning {#ctrl-pkt-congestion}

> The Congestion Warning control packet is not in use at the moment.

Author: It is a legacy packet. The sender never sends it. But the receiver has a known behavior upon reception of this packet. Therefore, to avoid potential backward incompatibility this description is beneficial.

If the receiver receives this packet, it slows down sending rate

> increasing the minimum inter packet sending interval by 12.5%.

Reviewer: I think it should be explained why `12.5%` is, but not for this commit.

Author: Magic number.

: A single value or a range of lost packets sequence numbers. See packet sequence number
coding in {{packet-seq-list-coding}}.

### Congestion Warning {#ctrl-pkt-congestion}

> The Congestion Warning control packet is not in use by SRT sender
> at the moment and reserved for the future.
> However, the receiver handles this type of packet if it ever arrives.
> 
> If the receiver receives this packet, it slows down sending rate
> increasing the minimum inter packet sending interval by 12.5%.

Reviewer 2: This description (1108 - 1113) sounds to me like it is implementation-specific. Maybe we can make it less so:

```suggestion
The Congestion Warning control packet is reserved for future use. Its purpose is to signal a receiver to increase the minimum inter-packet sending interval by a discrete value in order to slow down the sending rate.
```

Also, isn't it the sender that sets the minimum inter-packet sending interval?

Author: > Also, isn't it the sender that sets the minimum inter-packet sending interval?

Right!!! I mixed up "control packet receiver" and "receiver".
Should be: "If a sender receives this Congestion Warning..."

> The Congestion Warning control packet is reserved for future use.
> Its purpose is to signal a sender about a kind of congestion happening at the receiving side.
> 
> There is no condition to send this type of packet defined yet.
> However, if the sender receives this packet, it slows down sending rate
> increasing the minimum inter packet sending interval by a discrete value (12.5%).

Reviewer 2: ```suggestion
The Congestion Warning control packet is reserved for future use. 
Its purpose is to allow a receiver to signal a sender that there is congestion 
happening at the receiving side. The expected behaviour is that upon 
receiving this packet the sender slows down its sending rate by increasing the 
minimum inter-packet sending interval by a discrete value (posited to be 12.5%).

Note that the conditions for a receiver to issue this type of packet are not yet defined.

```

ACKACK control packets do not contain Control Information Field (CIF).

### Message Drop Request {#ctrl-pkt-dropreq}

The Message Drop Request is sent by the sender to the receiver
in case of some internal error, when an unacknowledged packet is not present
in sender's buffer. Thus the sender notifies the receiver that it must not
wait for retransmission of this message.
Note the Message Drop Request is not sent if sender decides to drop a message

> according to the TL Packet Drop mechanism, as in this case the receiver

Reviewer: "TL Packet Drop" first appeared here. For the term consistency, it would be better to be "Too-Late Packet Drop" or "TLPKTDROP".

ACKACK control packets do not contain Control Information Field (CIF).

### Message Drop Request {#ctrl-pkt-dropreq}

> The Message Drop Request is sent by the sender to the receiver
> in case of some internal error, when an unacknowledged packet is not present
> in sender's buffer. Thus the sender notifies the receiver that it must not
> wait for retransmission of this message.
> Note the Message Drop Request is not sent if sender decides to drop a message
> according to the Too Late Packet Drop mechanism ({{too-late-packet-drop}}),
> as in this case the receiver is expected to drop it anyway due to the same mechanism.

Reviewer 2: ```suggestion
A Message Drop Request control packet is sent by the sender to the receiver 
when it requests the retransmission of an unacknowledged packet (all or part 
of a message) which is not present in the sender's buffer (due to an internal error). 
The sender notifies the receiver that it must not wait for retransmission of this 
message. Note that a Message Drop Request control packet is not sent if the 
Too Late Packet Drop mechanism ({{too-late-packet-drop}}) causes the sender 
to drop a message, as in this case the receiver is expected to drop it anyway.
```

|                     Destination Socket ID                     |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                 First Packet Sequence Number                  |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                  Last Packet Sequence Number                  |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~
{: #dropreq-control-packet title="Drop Request control packet"}

Packet Type: 1 bit, value = 1.
: The packet type value of an ACKACK control packet is "1".

Control Type: 15 bits, value = 7.
: The control type value of a Drop Request control packet is "7".

Message Number: 32 bits.
: The identifying number of the message requested to be dropped.
  See the Message Number field in {{data-pkt}}.

Timestamp: 32 bits.
: See {{packet-structure}}.

Destination Socket ID: 32 bits.
: See {{packet-structure}}.

First Packet Sequence Number: 32 bits.
: The sequence number of the first packet in the message.

Last Packet Sequence Number: 32 bits.
: The sequence number of the last packet in the message.

### Peer Error {#ctrl-pkt-peer-error}

> This control packet is only used if file congestion control is enabled.
> The Peer Error control packet is signaled from the receiving side
> in case to some processing failures (e.g., write to disk has failed),
> to inform the sender about such situation and unblock it from waiting further
> response from the peer.

Reviewer 2: ```suggestion
The Peer Error control packet is sent by a receiver when a processing error 
(e.g. write to disk failure) occurs. This informs the sender of the situation 
and unblocks it from waiting for further responses from the receiver.

**NOTE**: This control packet is only used if file congestion control is enabled.

```

> The sender receiving this type of control packet must unblock any sending operation in progress.

Reviewer 2: ```suggestion
The sender receiving this type of control packet must unblock any sending operation in progress.

**NOTE**: This control packet is only used if file congestion control is enabled.
```

~~~
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+- SRT Header +-+-+-+-+-+-+-+-+-+-+-+-+-+
|1|      Control Type = 8       |         Reserved = 0          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                          Error Code                           |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                           Timestamp                           |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                     Destination Socket ID                     |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~
{: #peer-error-control-packet title="Peer Error control packet"}

Packet Type: 1 bit, value = 1.

> : The packet type value is "1".

Reviewer 2: ```suggestion
: The packet type value of a Peer Error control packet is "1".
```

Control Type: 15 bits, value = 8.

> : The control type value is "8".

Reviewer 2: ```suggestion
: The control type value of a Peer Error control packet is "8".
```
